### ISSUE TITLE
Internal server error: exceptions::mutation_write_timeout_exception——received only 1 responses from 2 CL=LOCAL_SERIAL.)

### ISSUE THREAD

**Issue by Felix-zhoux** [scylladb/scylla#6479]
*Installation details*
Scylla version (or git commit hash): 4.0.0-0.20200514.d95aa77b6
Cluster size:  9 nodes 
OS (RHEL/CentOS/Ubuntu/AWS AMI): CentOS

I used the aws-sdk-cpp developer tool to test Scylla through alternator, and I got a lot of HTTP 500 errors on the client side. When I opened the trace log of alternator-server, I found the following error:
```
DEBUG 2020-05-19 09:13:03,547 [shard 22] alternator-server - Skipping authorization
TRACE 2020-05-19 09:13:03,575 [shard 7] alternator-server - api_handler success case
TRACE 2020-05-19 09:13:03,583 [shard 11] alternator-server - api_handler success case
TRACE 2020-05-19 09:13:03,589 [shard 1] alternator-server - api_handler error case: {"__type":"com.amazonaws.dynamodb.v20120810#Internal Server Error","message":"Internal server error: exceptions::mutation_write_timeout_exception (Operation timed out for alternator_.dir.e441b8a6-187b-4b98-9c7a-2422c4f4fda4.4240.1..dir.e441b8a6-187b-4b98-9c7a-2422c4f4fda4.4240.1 - received only 1 responses from 2 CL=LOCAL_SERIAL.)"}
TRACE 2020-05-19 09:13:03,595 [shard 0] alternator-server - api_handler success case
TRACE 2020-05-19 09:13:03,601 [shard 10] alternator-server - api_handler success case
TRACE 2020-05-19 09:13:03,606 [shard 22] alternator-server - api_handler success case
```

KEYSPACE AND TABLE INFORMATION:
```
cqlsh> DESC KEYSPACE "alternator_.dir.e441b8a6-187b-4b98-9c7a-2422c4f4fda4.4240.1";

CREATE KEYSPACE "alternator_.dir.e441b8a6-187b-4b98-9c7a-2422c4f4fda4.4240.1" WITH replication = {'class': 'NetworkTopologyStrategy', 'dc1': '3', 'dc2': '3', 'dc3': '3'}  AND durable_writes = true;

CREATE TABLE "alternator_.dir.e441b8a6-187b-4b98-9c7a-2422c4f4fda4.4240.1".".dir.e441b8a6-187b-4b98-9c7a-2422c4f4fda4.4240.1" (
    bi text,
    obj text,
    ":attrs" map<text, blob>,
    PRIMARY KEY (bi, obj)
) WITH CLUSTERING ORDER BY (obj ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'SizeTieredCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';

scylla_tags = {}
```

**Comment by avikivity** [comment]
It may be that the workload is too intensive for the cluster. How many threads are running on the client, and what does the server hardware look like?

**Comment by Felix-zhoux** [comment]
> It may be that the workload is too intensive for the cluster. How many threads are running on the client, and what does the server hardware look like?

I ran three instances of Scylla on a physical node, compiled using frozen toolchain.The following is the hardware information of my physical node:

```
CPU  Intel(R) Xeon(R) CPU E5-2620 v4 @ 2.10GHz * 2
MemTotal:       131923628 kB
Intel® SSD D3-S4610 Series (480GB, 2.5in SATA 6Gb/s, 3D2, TLC) Generic Single Pack（one SSD per scylla instance）
```

16 threads were used in the alternator test. And I found that the total ops of the cluster at the time of testing was similar to what I had previously tested using the Cassandra-stress tool.Looking through the grafana panel, the Scylla cluster was not heavily loaded.

**Comment by Felix-zhoux** [comment]
![image](https://user-images.githubusercontent.com/59597542/82312595-e2406980-99f9-11ea-99eb-e5c4358e80a6.png)
![image](https://user-images.githubusercontent.com/59597542/82312914-46632d80-99fa-11ea-9250-5ebc90e6b214.png)

**Comment by nyh** [comment]
Grasping at straws here, but I can't help but notice that in your original report, I see trace-level log messages in the log. If you enable those, just logging all these silly messages can really tank performance. Can you please make sure they aren't enabled?

The error message mentions "LOCAL_SERIAL", which we use for LWT (which is used in Alternator writes, by default). The issue might be related to LWT. So I have two questions:
1. Does your workload write to the same items again and again a lot, or to different items?
2. Can you try the same thing without LWT (set the "forbid_rmw" write isolation policy on this table, as explained in docs/alternator/alternator.md - if this fits this workload) and see if the problem is indeed specific to our use of LWT?

**Comment by Felix-zhoux** [comment]
> Grasping at straws here, but I can't help but notice that in your original report, I see trace-level log messages in the log. If you enable those, just logging all these silly messages can really tank performance. Can you please make sure they aren't enabled?
> 
> The error message mentions "LOCAL_SERIAL", which we use for LWT (which is used in Alternator writes, by default). The issue might be related to LWT. So I have two questions:
> 
> 1. Does your workload write to the same items again and again a lot, or to different items?
> 2. Can you try the same thing without LWT (set the "forbid_rmw" write isolation policy on this table, as explained in docs/alternator/alternator.md - if this fits this workload) and see if the problem is indeed specific to our use of LWT?

To answer some of your questions first：
1. I opened the trace log manually to see what happened.
2. With LWT enabled by default, whether my workload writes different items or the same items, the client will get some 500 responses.
3. After I turned off the LWT feature and wrote different items, I found that there were no more 500 errors on the client side, and the delay of alternator's putitem was greatly reduced.

**Comment by Felix-zhoux** [comment]
I wonder why LWT can cause such a big delay, can we do some optimization for it?When LWT is enabled, the 99th percentile PutItem latency by Cluster in the alternator reaches 524ms. When LWT is turned off, the delay is only 2ms, which occurs when I write different items.

before
![image](https://user-images.githubusercontent.com/59597542/82851185-9cbbf900-9f31-11ea-9ef5-92da9fc6c39e.png)

after
![image](https://user-images.githubusercontent.com/59597542/82851128-5ff00200-9f31-11ea-83b5-0d97f5357ba4.png)

**Comment by slivne** [comment]
@Felix-zhoux if you share the metrics during the run with LWT + without tracing we can try to check if this is an overloaded case 

Pleas upload the metrics https://docs.scylladb.com/troubleshooting/report_scylla_problem/#server-metrics-with-prometheus

**Comment by Felix-zhoux** [comment]
> @Felix-zhoux if you share the metrics during the run with LWT + without tracing we can try to check if this is an overloaded case
> 
> Pleas upload the metrics https://docs.scylladb.com/troubleshooting/report_scylla_problem/#server-metrics-with-prometheus

@slivne  hi, I have uploaded the monitoring data of Prometheus according to the steps on the webpage, please check whether it is the data you want.

My report_uuid is:
`86803743-4568-4fac-8c33-6d94868e2d80`

**Comment by slivne** [comment]
@eliransin can you please have someone look at the metrics shared

**Comment by slivne** [comment]
@eliransin ping

**Comment by psarna** [comment]
Here are some LWT metrics extracted from the upload
![lwt](https://user-images.githubusercontent.com/10433047/89296392-8ed9d100-d662-11ea-947a-15fa6d7f9306.png)

**Comment by psarna** [comment]
LWT average write latency definitely correlates with PutItem latency, but I don't see a clear reason why it's the case... contention seems to be at quite low levels, unless I misread the metrics.

**Comment by psarna** [comment]
Interesting to note that only a single node (orange, 192.168.3.14) experiences such huge latencies, the rest is sub-100ms.

**Comment by psarna** [comment]
Some more metrics where the orange node stands out:
![cpu](https://user-images.githubusercontent.com/10433047/89298817-1f65e080-d666-11ea-9fc8-16fcb9f45d39.png)
![query_io](https://user-images.githubusercontent.com/10433047/89298823-20970d80-d666-11ea-8f14-e8f878f3dedd.png)

**Comment by psarna** [comment]
Maybe it got overloaded because it acted as coordinator for all client threads? That doesn't exactly explain 6s latencies for LWT, but maybe it's one of the root causes.

**Comment by slivne** [comment]
@gleb-cloudius / @kostja are there lwt metrics showing the info if a node had contention - if it was coordinator and for how many requests.

@psarna I thinkl we have additional lwt metrics - that can help you answer the questions you have

**Comment by gleb-cloudius** [comment]
On Sun, Sep 13, 2020 at 03:01:14AM -0700, Shlomi Livne wrote:
> @gleb-cloudius / @kostja are there lwt metrics showing the info if a node had contention - if it was coordinator and for how many requests.
> 
There are cas_read_contention/cas_write_contention. Not receiving a
response is not caused by contention though. Usually it means dead or
nearly dead node.

--
			Gleb.

**Comment by psarna** [comment]
The contention metrics are included in the first screenshot I posted, but they do not indicate any problems on this node.
Anyway, I took a closer look at the shard level, suspecting that maybe a single shard gets overloaded, and it looks like the suspicion was correct.
Here, only shard8 of the most overloaded node experiences large latencies:
![hot_shard](https://user-images.githubusercontent.com/10433047/93237841-f4040480-f780-11ea-8c4a-6eb4fcd1e1e5.png)
![hot_shard2](https://user-images.githubusercontent.com/10433047/93237851-f6fef500-f780-11ea-9bc1-410a90e18097.png)

And here's a closer look at the distribution of rows - it's clear that shard8 of each node is severely imbalanced, containing millions of rows, while other shards containing almost none.
![hot_shards3](https://user-images.githubusercontent.com/10433047/93237862-f9f9e580-f780-11ea-8949-f84951a78aaa.png)

That suggests a very imbalanced workload, so I guess that's why write timeout exceptions happened - one shard of each node was overloaded, and one node (probably the owner of the huge partition and/or the coordinator of every operation) was even more overloaded than the rest.
